docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,455 @@
+
+Additionally, each HelmReleaseProxy will include a label indicating the Cluster the release is installed on, and an additional label indicating the HelmChartProxy it corresponds to. This provides a way to query for the Helm releases installed on a Cluster and also the Helm releases installed from a HelmChartProxy. This implementation also generalizes to any ClusterAddonProvider as it uses a CRD to maintain the inventory instead of the package manager.
+
+#### Surface Cluster Add-ons Status
+
+ClusterAddonProviders are responsible for installing add-ons on an arbitrary number of workload clusters. As such, it would be beneficial to see the status of an add-on from the management cluster and avoid having to fetch the kubeconfig of the workload cluster.
+
+In the ClusterAddonProvider for Helm, the status of a Helm release can be conveyed through the status field of a HelmReleaseProxy. It includes a ready status, which conveys that the HelmRelease is installed and up to date with the CRD spec and a failure reason which is set if there is an error in reconciliation.
+
+To convey additional information, the HelmReleaseProxy status can include an array of `clusterv1.Conditions` and define new condition types as needed to offer more specific information. For example, there could be a `PodsReady` and `DeploymentsReady` condition.
+
+#### Upgrade strategy
+
+Users will be able to upgrade the add-ons across multiple Clusters at once since updating the version in the HelmChartProxy triggers a Helm upgrade for all the Helm releases it manages. Additionally, HelmChartProxy allows users to upgrade their add-ons in a declarative manner.
+
+Note that hooks and automatic triggers for add-on upgrades when a Cluster is upgraded to a new Kubernetes version is out of scope for this iteration. Users will need to determine if an add-on version upgrade is needed, and if so, set a compatible version in the HelmChartProxy.
+
+In future iterations, we plan to handle the upgrade of Cluster add-ons when the workload Cluster is upgraded to a new Kubernetes version by leveraging the new runtime hooks feature.
+
+#### Managing dependencies
+
+Managing dependencies between HelmChartProxies is out of scope for this project. However, Helm charts define their dependencies in the [Chart.yaml file](https://helm.sh/docs/topics/charts/#the-chartyaml-file) and will install them automatically. If a Helm chart has a critical dependency on another chart, it is the chart owner's responsibility to enable it by default.
+
+### Implementation details 
+
+The implementation will include two CRDs and two controllers as follows. As this is a prototype, the specific details regarding the CRD specs are subject to change.
+
+#### HelmChartProxy CRD
+
+The first CRD, HelmChartProxy, serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 
```

Comment:
I don't believe it's namespaced currently. HelmChartProxy should be able to select a Cluster across namespaces and do a Helm install since all it really needs is to fetch the spec and the kubeconfig and then initialize a Helm client. What would be the pros and cons of making it namespaced?

Tagging @elmiko since it's related to the namespace issue of the `clusterRef` in HelmReleaseProxy.

Reply:
cc @fabriziopandini

Reply:
Hm good question, not sure about the trade-offs. Currently all our CRDs except ExtensionConfig are namespaced.

Could have impact on `cluster move / backup / restore` (if it's in scope of that).

Maybe also relevant regarding RBAC if you want to give folks access to only one cluster (+ all its resouces) + it's addon CRs, that's probably impossible with a cluster-wide CRD.

On the other side of course it would limit the usefulness as the CRDs would probably only work with the Clusters in a single namespace.

Not sure if we have some historic background that could help here.

Reply:
One idea I'm fleshing out is to add a  `namespaces` field to the `clusterSelector` based on [this](https://capz.sigs.k8s.io/topics/multitenancy.html#allowednamespaces). The idea would be that you can

1. Default to the namespace of the HCP if it's not set
2. Pass in a list of namespaces to search for Clusters
3. Use something like a wildcard `*` to search across all namespaces 

WDYT about something along those lines?

Reply:
It highly depends on the CRDs being namespaced or cluster-wide. If they are namespaced, the CRs should probably only act on the CAPI CRDs in the same namespace (like other CAPI CRDs)

Reply:
Gotcha. I'd like to move forward with the CRDs being cluster-wide for ease of use (so you don't need 100 HelmChartProxies if you have 100 namespaces), but I'm open to making them namespaced if there's a strong case for doing so.